Support MsgCancelProposal and improve v0.5.0 compatibility

thismars opened the pull request:
# PR Description

This PR introduces several updates to ensure compatibility with Cosmos SDK v0.5.0 and adds key functionality as follows:

# Key Changes

### 1. Support for MsgCancelProposal
- Added Amino type declaration for MsgCancelProposal in firmajs
- Implemented unit tests to verify the expected behavior

### 2. Refactoring for v0.5.0 Compatibility
- Updated Amino type path for MsgAddContractLog:
   `/firmachain.firmachain.contract.MsgAddContractLog → /firmachain.contract.MsgAddContractLog`
- Revised timestamp conversion logic to include both seconds and nanos as required by v0.5.0

### 3. REST API Path Update
- Updated bank supply endpoint path:
   `/cosmos/bank/v1beta1/supply/${denom} → /cosmos/bank/v1beta1/supply/by_denom?denom=${denom}`
   (to address changes introduced by v0.5.0)

### 4. Test Enhancements
- Added utility function extractAllNftIds to extract NFT IDs from message events
- Fixed failing test scripts for consistent test execution
- Removed unnecessary imports and variables for cleaner code

# Additional Notes

- Addresses structural changes introduced by Cosmos SDK v0.5.0 (e.g., timestamp and REST API format updates)
- Improves test coverage and maintainability with utility enhancements

# Purpose

To ensure firmajs works reliably under Cosmos SDK v0.5.0 and to introduce support for the newly needed MsgCancelProposal message type.

# Summarize

- Total affected test cases: **311**
- Pass: **265**
- Skip: **41**
- Error: **5**

![image](https://github.com/user-attachments/assets/2df2dac4-3242-4e8f-ba89-3bb514bb5a48)

The following test results were produced by running npm test for firma-js against a locally configured devnet chain.
This summary includes passed, failed, and skipped tests to help evaluate the impact of this update.

## Test results

### Passed Tests (265)
- Out of 311 total tests, 265 passed successfully.
- Core functionalities (wallet, contract, bank, token transfer, authorization, etc.) worked as expected.
- See the screenshots below for actual test output results.

### Failed Tests (5)
- **[08. Gas Estimation Test]**
  5-2. Staking undelegate: max delegation count reached (passes with wallet change)
  5-3. Staking redelegate: max redelegation count reached (passes with wallet change)

- **[12. Staking Tx Test]**
  undelegate/redelegate: failed for the same reason as above (passes with wallet change)

- **[23. Authz Tx Test]**
  undelegate: max undelegation count reached (passes with wallet change)

### Skipped Tests (41)
- **[08. Gas Estimation Test]**
  7-2 ~ 7-7. Governance tests skipped due to typeUrl mismatch
  (Current: /cosmos.gov.v1beta1.MsgSubmitProposal, Expected: /cosmos.gov.v1.MsgSubmitProposal)

- **[12. Staking Tx Test]**
  editValidator: skipped due to 24-hour time restriction after last update

- **[16. Gov Tx Test]**
  SubmitCommunityPoolSpendProposal
  SubmitParameterChangeProposal
  SubmitSoftwareUpgradeProposalByHeight
  SubmitCancelSoftwareUpgradeProposal
  Vote - alice YES
  Vote - bob NO
  → All skipped due to typeUrl mismatch

- **[25. CosmWasm Tx Test]**
  MigrateContract: skipped due to unidentified error

- **[28. IBC Tx Test]**
  All tests skipped (no IBC node available on devnet)

- **[30. cw20 tx Test]**
  send
  send_from
  → skipped due to unidentified error

- **[31. cw20 query Test]**
  getDownloadLogo: skipped due to error when logo data is missing

- **[32. cw721 tx Test]**
  send_nft: due to unidentified error
  
- **[37. Marketplace tx Test]**
  All mint/register/purchase/cancel scenarios skipped
  → due to parameter mismatch between firma-js and contract

- **[37. Marketplace query Test]**
  get config
  get owner
  get register list
  → skipped for the same reason
  
### Test Output Screenshots
  
Wallet test output  
![image](https://github.com/user-attachments/assets/0164b94b-5299-4ff2-a154-cceea1a53c4d)

Contract test output  
![image](https://github.com/user-attachments/assets/04941067-7442-4310-bb97-4d7995407980)

Bank test output  
![image](https://github.com/user-attachments/assets/8c71b1ac-9872-4e16-9ef2-d7c0e7ca1da2)

Feegrant test output  
![image](https://github.com/user-attachments/assets/8cf375bc-8ebe-4312-82e3-eb0b2a4f908e)

GasEstimation test output  
![image](https://github.com/user-attachments/assets/c23cac68-e1bd-4fbc-ae61-2de34a778c98)

IPFS & NFT test output  
![image](https://github.com/user-attachments/assets/f6cb5632-986b-4e08-8631-fdb6f4047b6a)

Staking test output  
![image](https://github.com/user-attachments/assets/07973519-8395-44b6-b6c7-6dacfd6b4871)

Distribution test output  
![image](https://github.com/user-attachments/assets/738bd353-3a21-443c-a45c-06553a8d0560)

Gov test output  
![image](https://github.com/user-attachments/assets/5530606e-14d5-4901-a17c-7eacc30831b5)

Util & Chain & Slashing test output  
![image](https://github.com/user-attachments/assets/f4267c8b-f471-4480-8b23-6efb80fe1852)

Token test output  
![image](https://github.com/user-attachments/assets/befeb74c-cc4e-4887-a582-f69a9961d47b)

Authz test output  
![image](https://github.com/user-attachments/assets/0a94a7cb-a311-4de0-ab08-43dedfabba38)

Cosmwasm test output  
![image](https://github.com/user-attachments/assets/e95264a9-6eda-4b4b-9ee6-ead09ce22084)

Arbitary Sign test output  
![image](https://github.com/user-attachments/assets/fcb6290b-a2ed-454e-b1f1-7ad0605c7beb)

Mint test output  
![image](https://github.com/user-attachments/assets/bd6968d5-16ba-4045-baa5-6cbbda45f4de)

CW20 test output  
![image](https://github.com/user-attachments/assets/e6f04588-64dc-4e68-b5ef-d7b0d14b5087)

CW721 test output  
![image](https://github.com/user-attachments/assets/563236e9-16d3-4e71-aa36-38e10efce7e3)

CWBridge test output  
![image](https://github.com/user-attachments/assets/2ecf9674-ac82-40bf-a298-741c71db5f09)

CWMarket test output  
![image](https://github.com/user-attachments/assets/236f5991-9cbb-4a69-9187-17192c7b5913)

thismars committed:
feat: change type url to match v0.5.0

thismars committed:
fix: update bank supply endpoint to use query param format

thismars committed:
feat: add utility function to extract NFT IDs from message events

thismars committed:
fix: include nanos in timestamp

thismars committed:
refactor: clean imports and apply gas calculation

thismars committed:
feat: support direct sign TxRaw and enhance event parsing

thismars committed:
feat: msgcancelproposal preparation

thismars committed:
refactor: update cw721 and cwBridge tx/query scripts with runnable structure

thismars committed:
chore: remove unused imports and dead code

thismars committed:
feat: add MsgCancelProposal support and related functionality

- Added support for MsgCancelProposal in the governance module.
- Updated package dependencies to include @kintsugi-tech/cosmjs-types.
- Implemented gas estimation and transaction signing for canceling proposals.
- Enhanced tests to cover the new cancel proposal functionality.

koofree commented on test/06.feegrant_tx.test.ts:
```diff
@@ -28,13 +27,17 @@ describe('[06. Feegrant Tx Test]', () => {
-		expect(result.code).to.equal(0);
+		try {
+			var result = await firma.FeeGrant.grantPeriodicAllowance(aliceWallet, await bobWallet.getAddress(), periodicAllowanceData);
+			console.log(result);
```
I don't think this console.log is needed in here

thismars replied:
I will remove the log line.

koofree commented on test/06.feegrant_tx.test.ts:
```diff
@@ -28,13 +27,17 @@ describe('[06. Feegrant Tx Test]', () => {
 	});
 
-	it('feegrant RevokeAllowanec tx1', async () => {
+	it.skip('feegrant RevokeAllowanec tx1', async () => {
```
Why was this test skipped?

thismars replied:
Unnecessary unit test. I will delete it.

thismars replied:
removed .skip

https://github.com/FirmaChain/firma-js/pull/2/commits/e43e6e5d59468e1ce242ea3d56d2326c665ef383

koofree commented on test/10.nft_tx.test.ts:
```diff
@@ -18,18 +34,16 @@ describe('[10. NFT Tx Test]', () => {
-		var nftId = jsonData[0]["events"][0]["attributes"][2]["value"];
-
+		const nftId = extractAllNftIds(result.events);
+		// console.log(nftId);
```
It's better to remove console.log comment. (same to others)

thismars replied:
Checked it.

thismars replied:
removed console.log

https://github.com/FirmaChain/firma-js/pull/2/commits/a828b8826a3c252d7b382f6f40a3e71d89526af5

koofree commented on test/16.gov_tx.test.ts:
```diff
@@ -132,6 +152,47 @@ For a more detailed upgrade guide, please visit https://github.com/FirmaChain/ma
 	});
 
+	it.only('SubmitTextProposal & CancelProposal Test', async () => {
+
```
It's not a good use of the log-on test code. Use assert instead of it plz.

koofree commented on test/16.gov_tx.test.ts:
```diff
@@ -178,18 +239,6 @@ For a more detailed upgrade guide, please visit https://github.com/FirmaChain/ma
 		//console.log(result);
 		expect(result.code).to.equal(0);
-
-		const validator1 = await firma.Wallet.fromMnemonic("angry water bunker where iron absurd cruise deliver clutch unique creek pyramid arch express flush pill lens concert absent enemy boring mom nuclear rose");
```
previous test codes already has mnemonics.. it's good to remove it

haryunio replied:
it's been a bad convention for a long time; will remove all of them

koofree commented on test/21.token_tx.test.ts:
```diff
@@ -1,12 +1,27 @@
 import { FirmaSDK } from "../sdk/FirmaSDK"
 import { aliceMnemonic, bobMnemonic, TestChainConfig } from './config_test';
 
 describe('[21. Token Tx Test]', () => {
```
Can you check the convention of `'` or `"`?

thismars replied:
I confirmed the preferred quote convention (' vs ") earlier, and have reflected that in the following commit.

https://github.com/FirmaChain/firma-js/pull/2/commits/28c105cff2b15ecce36b78e84db416fc511f4e55

koofree commented on test/25.cosmwasm_tx.test.ts:
```diff
@@ -36,17 +49,12 @@ describe('[25. CosmWasm Tx Test]', () => {
-
-		codeId = data[0]["events"][1]["attributes"][1]["value"];
+		const result = await firma.CosmWasm.storeCode(aliceWallet, array, everyBodyAccessConfig, { gas: gas, fee: fee });
+		codeId = extractValue(result.events, "store_code", "code_id");
```
codeId is not used for this.

thismars replied:
codeId used next unit test.

`const result = await firma.CosmWasm.instantiateContract(aliceWallet, admin, codeId, label, testData, funds, { gas: gas, fee: fee });`

koofree commented on test/27.arbitary_sign.test.ts:
```diff
@@ -63,36 +62,27 @@ describe('[27. arbitary sign]', () => {
 
-		let signDoc = FirmaUtil.parseSignDocValues(stringSignDoc);
+		// 3. Verify the signature against the original signDoc
+    const signatureBase64 = FirmaUtil.arrayBufferToBase64(txRaw.signatures[0]);
```
Different line tab.

thismars replied:
line tab spacing

https://github.com/FirmaChain/firma-js/pull/2/commits/d0714cbdbd8b6a9db3aba22ea98165263a600bbf

haryunio replied:
lgtm

koofree commented:
Can you let me know how to test this PR on my local labtop? @thismars

thismars committed:
style: unify string quotes from " to '

thismars committed:
test: remove .skip to re-enable unit test

thismars committed:
refactor: remove debug logs and tidy up blank lines

thismars committed:
style: fix inconsistent indentation and line tab spacing

thismars committed:
test: adjust skip status on specific tests

koofree commented on test/05.bank_query.test.ts:
```diff
@@ -12,78 +12,52 @@ describe('[05. Bank test]', () => {
-
-		//expect(result).to.be.equal("0");
+		const wallet = await firma.Wallet.fromMnemonic(firmaFeeMnemonic);
+		const result = await firma.Bank.getBalance(await wallet.getAddress());
```
`expect ' has disappeared. Does this test only check that it works without error?

koofree commented on test/33.cw721_query.test.ts:
```diff
@@ -1,22 +1,32 @@
 
 		const tokenId = "1";
 		const owner = await firma.Cw721.getOwnerFromNftID(contractAddress, tokenId);
-		console.log(owner);
```
I think we need not only to remove the console log but also to add a check of the result of the test with asserts.

thismars committed:
refactor: cleaned up logs and clarified return values

- Removed unnecessary commented-out log lines and clarified return values.
- Adjusted blank lines to improve code readability.

thismars committed:
test: Add additional test cases and improve validation logic with more assertions

thismars commented:
I left a commit with your requirements.
So please review.

https://github.com/FirmaChain/firma-js/pull/2/commits/0e9ca41f9fe88ccd174137b429a0a4f2df56d579

koofree commented on sdk/FirmaFeeGrantService.ts:
```diff
@@ -179,7 +179,7 @@ export class FirmaFeeGrantService {
 
             //console.log('Any.fromJSON(allowanceAnyData)');
-            //console.log(Any.fromJSON(allowanceAnyData));
+            //console.log('Any.fromJSON(allowanceAnyData)');
```
Is this `console.log` necessary for this service?

thismars replied:
Not important. I'll remove it.

thismars replied:
Please review the commit.
https://github.com/FirmaChain/firma-js/pull/2/commits/9266126672489cc850eb0f869119f22972f4c8af

haryunio replied:
lgtm

koofree commented on test/04.bank_tx.test.ts:
```diff
@@ -29,29 +27,26 @@ describe.only('[04. Bank Tx Test]', () => {
-			var result = await firma.Bank.send(wallet, await targetWallet.getAddress(), amount, { memo: "memo" });
-
+			await firma.Bank.send(wallet, await targetWallet.getAddress(), amount, { gas, fee });
+			expect(true).to.not.equal(true);
```
Always expect failure on this test code.

thismars replied:
I'll update the test code to make it clearer.

thismars replied:
Please review the commit.
https://github.com/FirmaChain/firma-js/pull/2/commits/9266126672489cc850eb0f869119f22972f4c8af

haryunio replied:
LGTM. Clear fail expectation and explicit message

koofree commented on test/06.feegrant_tx.test.ts:
```diff
@@ -28,63 +34,39 @@ describe('[06. Feegrant Tx Test]', () => {
-		var result = await firma.FeeGrant.grantBasicAllowance(aliceWallet, await bobWallet.getAddress());
-
-		expect(result.code).to.equal(0);
+		// var spendLimit = 200000;
```
remove unuse logs

thismars replied:
I'll remove it.

thismars replied:
Please review the commit.
https://github.com/FirmaChain/firma-js/pull/2/commits/9266126672489cc850eb0f869119f22972f4c8af

haryunio replied:
why some of feegrant tests are on the comment?

koofree commented on test/04.bank_tx.test.ts:
```diff
@@ -29,29 +27,26 @@ describe.only('[04. Bank Tx Test]', () => {
 
 		var result = await firma.Bank.send(wallet, await targetWallet.getAddress(), amount);
-
-		// error code 5 : not enough money.
```
Explanation comment is not needed? I think it could be helpful.

thismars replied:
I'll add a short explanation for clarity.

thismars replied:
Please review the commit.
https://github.com/FirmaChain/firma-js/pull/2/commits/9266126672489cc850eb0f869119f22972f4c8af

haryunio replied:
lgtm, clear explanation on code 5

koofree commented on sdk/firmachain/staking/StakingQueryClient.ts:
```diff
@@ -109,7 +109,7 @@ export class StakingQueryClient {
 
         // If there is no data in the list, throw 404 exception.
-        //console.log(result);
+        // console.log(result);
```
remove unuse log

thismars replied:
Please review the commit.
https://github.com/FirmaChain/firma-js/pull/2/commits/9266126672489cc850eb0f869119f22972f4c8af

haryunio replied:
lgtm

koofree commented:
gm @thismars, can you contain the captures of the test result in the PR?

thismars commented:
> gm @thismars, can you contain the captures of the test result in the PR?

Got it. I'll capture the test results and share them after running the tests.

thismars committed:
refactor: removed unnecessary comments and cleaned up code

thismars committed:
Refactor: Code cleanup and test improvements (00~03)
- Updated to reuse wallet addresses.

thismars committed:
Refactor: Code cleanup and test improvements (bank module)
- Updated to reuse wallet addresses.

thismars committed:
test: Code cleanup and test improvements (feegrant module)
- Updated to reuse wallet addresses.
- Reordered test cases (moved revoke test to the top)

koofree commented:
> > gm @thismars, can you contain the captures of the test result in the PR?
> 
> Got it. I'll capture the test results and share them after running the tests.

where can I get your test result?

thismars committed:
Refactor: Code cleanup and test improvements (gas test)
- Gov test requires core source modification (to accommodate typeUrl changes)

thismars committed:
refactor:Renaming variables for better readability

thismars committed:
test: Code cleanup and test improvements (nft module)
- Updated to reuse wallet addresses.

thismars commented:
> > > gm @thismars, can you contain the captures of the test result in the PR?
> > 
> > 
> > Got it. I'll capture the test results and share them after running the tests.
> 
> where can I get your test result?

I am currently improving the code to ensure all tests run successfully. I will let you know once all improvements are complete.

thismars committed:
test: Code cleanup and test improvements (staking module)

thismars committed:
test: Code cleanup and test improvements (distribution module)

thismars committed:
test: code cleanup and test improvements
– Update typeUrl to match the chain version (v1 → v1beta1)

thismars committed:
style: remove outdated comments

thismars committed:
test: code cleanup and test improvements (token module)
– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (authz module)

– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (cosmwasm module)

– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (arbitary sign module)

– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (ibc module)

– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (cosmwasm cw721)

– Simplified tests by reusing wallet addresses

thismars committed:
test: code cleanup and test improvements (cosmwasm cw bridge)

– Simplified tests by reusing wallet addresses

thismars committed:
test: skip cw market test

- update to be applied later

thismars committed:
test: modified to enable query testing (contract module)

thismars committed:
test: moved revoke position to top for stable test environment (feegrant)

thismars committed:
test: remove skip

thismars committed:
test: code cleanup and test improvements (cosmwasm cw20)

– Simplified tests by reusing wallet addresses

haryunio commented:
How are we going to deal with the governance patches? Separate branch, or a new PR?

koofree reviewed (approved):
I have checked a lot. It's too large a PR, so code review is no longer working effectively. I approve it. We need to find an alternative way to check the code, such as checking test results.

I approved by the testing result right now.

koofree commented:
> How are we going to deal with the governance patches? Separate branch, or a new PR?

I hope to do this work as a new PR. @haryunio @thismars
